blocks/table/src/components/table/table-with-query.tsx

```diff
@@ -0,0 +1,184 @@
+import {
+  Entity,
+  GraphBlockHandler,
+  JsonValue,
+  MultiFilter,
+} from "@blockprotocol/graph";
+import { getRoots } from "@blockprotocol/graph/stdlib";
+import {
+  DataEditorProps,
+  DataEditorRef,
+  GridCellKind,
+  GridColumn,
+} from "@glideapps/glide-data-grid";
+import { useEffect, useMemo, useRef, useState } from "react";
+
+import { RootKey } from "../../additional-types";
+import { BlockEntity } from "../../types/generated/block-entity";
+import { Grid, ROW_HEIGHT } from "../grid/grid";
+
+const isStripedKey: RootKey =
+  "https://blockprotocol-fwu7vped4.stage.hash.ai/@yk_hash/types/property-type/table-rows-are-striped/";
+const hideHeaderRowKey: RootKey =
+  "https://blockprotocol-fwu7vped4.stage.hash.ai/@yk_hash/types/property-type/table-header-row-is-hidden/";
+const hideRowNumbersKey: RootKey =
+  "https://blockprotocol-fwu7vped4.stage.hash.ai/@yk_hash/types/property-type/table-row-numbers-are-hidden/";
+
+interface TableProps {
+  blockEntity: BlockEntity;
+  readonly?: boolean;
+  query: MultiFilter;
+  graphModule: GraphBlockHandler;
+}
+
+export const TableWithQuery = ({
+  blockEntity,
+  readonly,
+  query,
+  graphModule,
+}: TableProps) => {
+  const gridRef = useRef<DataEditorRef>(null);
+
+  const {
+    properties: {
+      [isStripedKey]: isStriped = false,
+      [hideHeaderRowKey]: hideHeaderRow = false,
+      [hideRowNumbersKey]: hideRowNumbers = false,
+    },
+  } = blockEntity;
+
+  const [loading, setLoading] = useState(true);
+  const [entities, setEntities] = useState<Entity[]>([]);
+
+  useEffect(() => {
+    const init = async () => {
+      const res = await graphModule.queryEntities({
+        data: {
+          operation: { multiFilter: query },
+          graphResolveDepths: {
+            inheritsFrom: { outgoing: 0 },
+            constrainsValuesOn: { outgoing: 0 },
+            constrainsPropertiesOn: { outgoing: 0 },
+            constrainsLinksOn: { outgoing: 0 },
+            constrainsLinkDestinationsOn: { outgoing: 0 },
+            isOfType: { outgoing: 0 },
+            hasLeftEntity: { incoming: 0, outgoing: 0 },
+            hasRightEntity: { incoming: 0, outgoing: 0 },
+          },
+        },
+      });
+
+      if (!res.data) {
+        throw new Error(res.errors?.[0]?.message ?? "Unknown error");
+      }
+
+      const roots = getRoots(res.data.results);
+      setLoading(false);
+
+      setEntities(roots);
+    };
+
+    void init();
+  }, [graphModule, query]);
+
+  const columns = useMemo<GridColumn[]>(() => {
+    const uniqueEntityTypeIds = new Set<string>();
+
+    for (const entity of entities) {
+      for (const key of Object.keys(entity.properties)) {
+        uniqueEntityTypeIds.add(key);
+      }
+    }
```

Comment:
I don't know if this is better but it's an alternative
```suggestion
    const uniqueEntityTypeIds = new Set<string>(
      entities.flatMap(({ properties }) => Object.keys(properties)),
    );
```

Reply:
Looks better, applied here https://github.com/hashintel/hash/pull/2491/commits/b28ba382b2cf3ae85dfa76c6dd11d4023662c7c7